// wdio-vscode-service/dist/locators/1.47.0.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScmView = exports.CustomTreeItem = exports.locatorVersion = void 0;
const _1_46_0_1 = require("./1.46.0");
__exportStar(require("./1.46.0"), exports);
exports.locatorVersion = '1.47.0';
exports.CustomTreeItem = {
    ..._1_46_0_1.CustomTreeItem,
    elem: (label) => `.//div[@role='treeitem' and .//span[text()='${label}']]`
};
exports.ScmView = {
    ..._1_46_0_1.ScmView,
    changes: './/div[@role="treeitem" and .//div/text()="Changes"]',
    stagedChanges: './/div[@role="treeitem" and .//div/text()="Staged Changes"]',
    providerTitle: 'name',
    providerType: 'description'
};
