// @appium/images-plugin/build/lib/compare.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MATCH_TEMPLATE_MODE = exports.MATCH_FEATURES_MODE = exports.GET_SIMILARITY_MODE = exports.DEFAULT_MATCH_THRESHOLD = void 0;
exports.compareImages = compareImages;

require("source-map-support/register");

var _lodash = _interopRequireDefault(require("lodash"));

var _driver = require("appium/driver");

var _opencv = require("@appium/opencv");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const MATCH_FEATURES_MODE = 'matchFeatures';
exports.MATCH_FEATURES_MODE = MATCH_FEATURES_MODE;
const GET_SIMILARITY_MODE = 'getSimilarity';
exports.GET_SIMILARITY_MODE = GET_SIMILARITY_MODE;
const MATCH_TEMPLATE_MODE = 'matchTemplate';
exports.MATCH_TEMPLATE_MODE = MATCH_TEMPLATE_MODE;
const DEFAULT_MATCH_THRESHOLD = 0.4;
exports.DEFAULT_MATCH_THRESHOLD = DEFAULT_MATCH_THRESHOLD;

async function compareImages(mode, firstImage, secondImage, options = {}) {
  const img1 = Buffer.from(firstImage, 'base64');
  const img2 = Buffer.from(secondImage, 'base64');
  let result = null;

  switch (_lodash.default.toLower(mode)) {
    case MATCH_FEATURES_MODE.toLowerCase():
      try {
        result = await (0, _opencv.getImagesMatches)(img1, img2, options);
      } catch (err) {
        result = {
          count: 0
        };
      }

      break;

    case GET_SIMILARITY_MODE.toLowerCase():
      result = await (0, _opencv.getImagesSimilarity)(img1, img2, options);
      break;

    case MATCH_TEMPLATE_MODE.toLowerCase():
      result = await (0, _opencv.getImageOccurrence)(img1, img2, options);

      if (options.multiple) {
        return result.multiple.map(convertVisualizationToBase64);
      }

      break;

    default:
      throw new _driver.errors.InvalidArgumentError(`'${mode}' images comparison mode is unknown. ` + `Only ${JSON.stringify([MATCH_FEATURES_MODE, GET_SIMILARITY_MODE, MATCH_TEMPLATE_MODE])} modes are supported.`);
  }

  return convertVisualizationToBase64(result);
}

function convertVisualizationToBase64(element) {
  if (!_lodash.default.isEmpty(element.visualization)) {
    element.visualization = element.visualization.toString('base64');
  }

  return element;
}
