// @appium/images-plugin/build/lib/plugin.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.ImageElementPlugin = exports.IMAGE_STRATEGY = void 0;
exports.getImgElFromArgs = getImgElFromArgs;

require("source-map-support/register");

var _lodash = _interopRequireDefault(require("lodash"));

var _driver = require("appium/driver");

var _plugin = _interopRequireDefault(require("appium/plugin"));

var _compare = require("./compare");

var _finder = _interopRequireDefault(require("./finder"));

var _imageElement = require("./image-element");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const IMAGE_STRATEGY = '-image';
exports.IMAGE_STRATEGY = IMAGE_STRATEGY;

function getImgElFromArgs(args) {
  for (let arg of args) {
    if (_lodash.default.isString(arg) && arg.startsWith(_imageElement.IMAGE_ELEMENT_PREFIX)) {
      return arg;
    }
  }
}

class ImageElementPlugin extends _plugin.default {
  constructor(pluginName) {
    super(pluginName);
    this.finder = new _finder.default();
  }

  static newMethodMap = {
    '/session/:sessionId/appium/compare_images': {
      POST: {
        command: 'compareImages',
        payloadParams: {
          required: ['mode', 'firstImage', 'secondImage'],
          optional: ['options']
        },
        neverProxy: true
      }
    }
  };

  async compareImages(next, driver, ...args) {
    return await (0, _compare.compareImages)(...args);
  }

  async findElement(next, driver, ...args) {
    return await this._find(false, next, driver, ...args);
  }

  async findElements(next, driver, ...args) {
    return await this._find(true, next, driver, ...args);
  }

  async _find(multiple, next, driver, ...args) {
    const [strategy, selector] = args;

    if (strategy !== IMAGE_STRATEGY) {
      return await next();
    }

    this.finder.setDriver(driver);
    return await this.finder.findByImage(selector, {
      multiple
    });
  }

  async handle(next, driver, cmdName, ...args) {
    const imgElId = getImgElFromArgs(args);

    if (imgElId) {
      if (!this.finder.imgElCache.has(imgElId)) {
        throw new _driver.errors.NoSuchElementError();
      }

      const imgEl = this.finder.imgElCache.get(imgElId);
      return await _imageElement.ImageElement.execute(driver, imgEl, cmdName, ...args);
    }

    return await next();
  }

}

exports.ImageElementPlugin = exports.default = ImageElementPlugin;
