// wdio-vscode-service/dist/constants.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_PROXY_OPTIONS = exports.CMD_KEY = exports.DEFAULT_CACHE_PATH = exports.DEFAULT_VSCODE_WEB_PORT = exports.DEFAULT_VSCODE_WEB_HOSTNAME = exports.CHROMEDRIVER_DOWNLOAD_PATH = exports.CHROMEDRIVER_RELEASES = exports.VSCODE_WEB_STANDALONE = exports.VSCODE_MANIFEST_URL = exports.VSCODE_RELEASES = exports.DEFAULT_VSCODE_SETTINGS = exports.SETTINGS_KEY = exports.VSCODE_APPLICATION_ARGS = exports.DEFAULT_CHANNEL = exports.VSCODE_CAPABILITY_KEY = void 0;
const path_1 = __importDefault(require("path"));
exports.VSCODE_CAPABILITY_KEY = 'wdio:vscodeOptions';
exports.DEFAULT_CHANNEL = 'stable';
exports.VSCODE_APPLICATION_ARGS = {
    // https://github.com/microsoft/vscode/issues/84238
    noSandbox: true,
    // https://github.com/microsoft/vscode-test/issues/120
    disableUpdates: true,
    skipWelcome: true,
    skipReleaseNotes: true,
    disableWorkspaceTrust: true,
    disableExtensions: true
};
exports.SETTINGS_KEY = 'wdioVSCodeService.proxy';
exports.DEFAULT_VSCODE_SETTINGS = {
    'window.titleBarStyle': 'custom',
    'workbench.editor.enablePreview': false,
    'window.restoreFullscreen': true,
    'window.newWindowDimensions': 'maximized',
    'security.workspace.trust.enabled': false,
    'window.dialogStyle': 'custom',
    'files.simpleDialog.enable': true,
    /**
     * for service specific configuration of the VSCode API proxy
     */
    [exports.SETTINGS_KEY]: {}
};
exports.VSCODE_RELEASES = 'https://update.code.visualstudio.com/api/releases/stable';
exports.VSCODE_MANIFEST_URL = 'https://raw.githubusercontent.com/microsoft/vscode/%s/cgmanifest.json';
exports.VSCODE_WEB_STANDALONE = 'https://update.code.visualstudio.com/api/update/web-standalone/%s/latest';
exports.CHROMEDRIVER_RELEASES = 'https://chromedriver.storage.googleapis.com/LATEST_RELEASE_%s';
exports.CHROMEDRIVER_DOWNLOAD_PATH = 'https://chromedriver.storage.googleapis.com/%s/chromedriver_%s.zip';
exports.DEFAULT_VSCODE_WEB_HOSTNAME = 'localhost';
exports.DEFAULT_VSCODE_WEB_PORT = 3000;
exports.DEFAULT_CACHE_PATH = path_1.default.join(process.cwd(), '.wdio-vscode-service');
exports.CMD_KEY = process.platform === 'darwin' ? 'Meta' : 'Control';
exports.DEFAULT_PROXY_OPTIONS = {
    enable: true,
    port: undefined,
    commandTimeout: 60 * 1000,
    connectionTimeout: 5 * 60 * 1000
};
